src/lib.rs

```diff
@@ -9,84 +9,129 @@
 // except according to those terms.
 
 //! Utilities for random number generation
+//! 
+//! ## Example
+//! 
+//! ```rust
+//! // Rng is the main trait and needs to be imported:
+//! use rand::{Rng, thread_rng};
+//!
+//! // thread_rng is often the most convenient source of randomness:
+//! let mut rng = rand::thread_rng();
+//! if rng.gen() { // random bool
+//!     let x: f64 = rng.gen(); // random number in range (0, 1)
+//!     println!("x is: {}");
+//!     println!("Number from 0 to 9: {}", rng.gen_range(0, 10));
+//! }
+//! ```
 //!
 //! The key function is `Rng::gen()`. It is polymorphic and so can be used to
-//! generate any type supporting the [`Uniform`] distribution (i.e. `T` where
-//! `Uniform`: `Distribution<T>`). Type inference means that often a simple call
-//! to `rng.gen()` will suffice, but sometimes an annotation is required, e.g.
+//! generate many types; the [`Uniform`] distribution carries the
+//! implementations. In some cases type annotation is required, e.g.
 //! `rng.gen::<f64>()`.
 //!
-//! See the `distributions` submodule for sampling random numbers from
-//! distributions like normal and exponential.
-//!
-//! # Usage
+//! # Getting random values
 //!
-//! This crate is [on crates.io](https://crates.io/crates/rand) and can be
-//! used by adding `rand` to the dependencies in your project's `Cargo.toml`.
-//!
-//! ```toml
-//! [dependencies]
-//! rand = "0.4"
+//! The most convenient source of randomness is likely `thread_rng()`, which
+//! automatically initialises a fast algorithmic generator on first use per
+//! thread with thread-local storage.
+//! 
+//! If one wants to obtain random data directly from an external source it is
+//! recommended to use `EntropyRng` which manages multiple available sources
+//! or `OsRng` which retrieves random data directly from the OS. It should be
+//! noted that this is significantly slower than using a local generator like
+//! `thread_rng` and potentially much slower if `EntropyRng` must fall back to
+//! `JitterRng` as a source.
+//! 
+//! It is also common to use an algorithmic generator in local memory; for
```

Comment:
I think it is better to have this one as the second paragraph, and to reword a bit.

It is also quite common, and can have advantages, to use an algorithmic generator in local memory.
`StdRng` and `SmallRng` provide to generators selected for common use. `StdRng` can be recommended for all uses, the main advantages of `SmallRng` are much smaller state and faster initialisation while its disadvantage is that it may be easy for an attacker to predict; actual performance varies.
...

Reply:
I don't want to make the doc too long however — I'll try to find a compromise.